RFC-002-mapeo-protocol.md

```diff
@@ -206,13 +210,23 @@ compatible version.
 An implementor should only use the major (`N.x.x`) version number. A minor
 (`x.N.x`) and patch (`x.x.N`) version are included for possible future use.
 
-## 3.4 `Ping()`
+## 3.4 `Heartbeat()`
 
-A peer should sent periodic `Ping` messages in order to ensure that the
+A peer should sent periodic `Heartbeat` messages in order to ensure that the
 connection is still active. TCP does not do this automatically, and it can take
```

Comment:
I think these would serve the purpose of both a [keepalive](https://datprotocol.github.io/how-dat-works/#keepalive) and a heartbeat. Before we open the hypercore connection we need a keepalive to avoid a "network middlebox" killing the connection if no data is sent for a while (30 seconds+).

For the heartbeat (this may be a complexity to keep for a later iteration of this) I was wondering how to handle slow connections where the response might be 30 seconds+ (we did see this in Guyana quite a lot, although it is rare) but still be able to detect an actual broken connection reasonably quickly. Perhaps the heartbeat response timeout could start at a high number like 60 seconds, and then adjust based on a rolling average of heartbeat responses during the connection e.g. `avgResp = prevAvgResp + (lastResponseTime - prevAvgResp) / totalHeartbeats)`. Probably better to use [rolling variance](https://github.com/digidem/react-mapfilter/blob/v3/src/lib/data_analysis/statistics.js#L265) too to calculate a dynamic timeout.

Reply:
Ooh, good point. I think the simple initial implementation should be ok for now, and we can revisit a more latency-tolerant version later. Right now hypercore is already timing out at 20s, so these changes I'm proposing won't make Mapeo's current behaviour any *worse*.